libs/vkd3d/command.c

```diff
@@ -9988,6 +9988,70 @@ static void STDMETHODCALLTYPE d3d12_command_list_SetPredication(d3d12_command_li
     }
 }
 
+// *PIXEvent* enum values and consts are from PIXEventsCommon.h of winpixeventruntime package
+
+static char* decode_pix_blob(const void* data, size_t size)
+{
+    static const UINT64 PIXEventsStringIsANSIReadMask = 0x0040000000000000;
+    static const UINT64 PIXEventsTypeReadMask = 0x00000000000FFC00;
+    static const UINT64 PIXEventsTypeBitShift = 10;
+    UINT64* data_uint64_aligned = (UINT64*)(data);
+    size_t label_str_length;
+    char *label_str = NULL;
+    bool is_ansi;
+    UINT64 type;
+
+    enum PIXEventType
+    {
+        ePIXEvent_EndEvent = 0x000,
+        ePIXEvent_BeginEvent_VarArgs = 0x001,
+        ePIXEvent_BeginEvent_NoArgs = 0x002,
+        ePIXEvent_SetMarker_VarArgs = 0x007,
+        ePIXEvent_SetMarker_NoArgs = 0x008,
+
+        ePIXEvent_EndEvent_OnContext = 0x010,
+        ePIXEvent_BeginEvent_OnContext_VarArgs = 0x011,
+        ePIXEvent_BeginEvent_OnContext_NoArgs = 0x012,
+        ePIXEvent_SetMarker_OnContext_VarArgs = 0x017,
+        ePIXEvent_SetMarker_OnContext_NoArgs = 0x018,
+    };
+
+    type = (*data_uint64_aligned & PIXEventsTypeReadMask) >> PIXEventsTypeBitShift;
+
+    if (type == ePIXEvent_SetMarker_NoArgs)
+        type = ePIXEvent_BeginEvent_NoArgs;
+
+    if (type != ePIXEvent_BeginEvent_NoArgs)
+    {
+        WARN("Unexpected/unsupported PIX3Event type, SetMarker_VarArgs and SetMarker_VarArgs can be added on request");
+        return NULL;
+    }
+
+    data_uint64_aligned++; //skip Color
+    data_uint64_aligned++;
+
+    is_ansi = *data_uint64_aligned & PIXEventsStringIsANSIReadMask;
+
+    data_uint64_aligned++;
+    
+    if (is_ansi)
+    {
+        label_str_length = strlen((const char*)data_uint64_aligned);
+        label_str = vkd3d_malloc(label_str_length + 1);
+        if (!label_str)
+            return NULL;
+        memcpy(label_str, (const char*)data_uint64_aligned, label_str_length);
+        label_str[label_str_length] = '\0';
+    }
+    else
+    {
+        label_str_length = wcslen((const wchar_t*)data_uint64_aligned);
+        label_str = vkd3d_strdup_w_utf8((const wchar_t*)data_uint64_aligned, label_str_length);
```

Comment:
Deduce the length based on `size`, subtracted by the blob offset. Also it's `WCHAR` not `wchar_t`. The C type is 32-bit on Linux, but we explicitly need 16-bit.